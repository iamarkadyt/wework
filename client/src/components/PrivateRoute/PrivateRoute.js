import React from 'react'
import { Route, Redirect, withRouter } from 'react-router-dom'
import { connect } from 'react-redux'
import { oneOfType, node, func, shape, bool } from 'prop-types'

function privateRoute({ 
  component: Component, 
  authedUser: { 
    isAuthenticated 
  }, 
  ...rest 
}) {
  return (
    <Route {...rest} render={props =>
      isAuthenticated
        ? <Component {...props} />
        : <Redirect to='/login' />} />
  )
}

privateRoute.propTypes = {
  component: oneOfType([node, func]).isRequired,
  authedUser: shape({
    isAuthenticated: bool.isRequired
  }).isRequired
}

export { privateRoute as _UnconnectedPrivateRoute }
export default withRouter(connect(state => ({
    authedUser: state.user
}))(privateRoute))
