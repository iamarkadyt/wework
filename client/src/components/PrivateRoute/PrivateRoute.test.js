import React from 'react'
import { mount } from 'enzyme'
import { StaticRouter as Router } from 'react-router'
import { _UnconnectedPrivateRoute as PrivateRoute, mapStateToProps } from './PrivateRoute'
import cloneDeep from 'lodash.clonedeep'

let mountedComponent, props

const getMockProps = () => {
  return cloneDeep({
    component: () => <h1>Hello!</h1>,
    authedUser: { isAuthenticated: true }
  })
}

const comp = () => {
  if (!mountedComponent) {
    mountedComponent = mount(
      <Router context={{}}>
        <PrivateRoute exact {...props} />
      </Router>
    )
  }
  return mountedComponent
}

describe('PrivateRoute', () => {
  beforeEach(() => {
    mountedComponent = undefined
    props = getMockProps()
  })

  describe('props', () => {
    describe('effects', () => {
      describe('isAuthenticated', () => {
        it('if true, component is rendered', () => {
          props.authedUser.isAuthenticated = true
          expect(comp().find('component').exists()).toBe(true)
        })

        it('if false, Redirect is rendered', () => {
          props.authedUser.isAuthenticated = false
          expect(comp().find('Redirect').exists()).toBe(true)
        })
      })
    })
  })

  it('matches snapshot when isAuthenticated is true', () => {
    props.authedUser.isAuthenticated = true
    expect(comp()).toMatchSnapshot()
  })

  it('matches snapshot when isAuthenticated is false', () => {
    props.authedUser.isAuthenticated = false
    expect(comp()).toMatchSnapshot()
  })

  describe('mapStateToProps', () => {
    it('returns expected object', () => {
      const state = { user: 'name' }
      expect(mapStateToProps(state)).toEqual({ authedUser: 'name' })
    })
  })
})

