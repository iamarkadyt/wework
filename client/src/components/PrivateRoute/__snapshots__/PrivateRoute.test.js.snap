// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PrivateRoute matches snapshot when isAuthenticated is false 1`] = `
<StaticRouter
  basename=""
  context={
    Object {
      "action": "REPLACE",
      "location": Object {
        "hash": "",
        "pathname": "/login",
        "search": "",
        "state": undefined,
      },
      "url": "/login",
    }
  }
  location="/"
>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <privateRoute
      authedUser={
        Object {
          "isAuthenticated": false,
        }
      }
      component={[Function]}
      exact={true}
    >
      <Route
        exact={true}
        render={[Function]}
      >
        <Redirect
          push={false}
          to="/login"
        />
      </Route>
    </privateRoute>
  </Router>
</StaticRouter>
`;

exports[`PrivateRoute matches snapshot when isAuthenticated is true 1`] = `
<StaticRouter
  basename=""
  context={Object {}}
  location="/"
>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <privateRoute
      authedUser={
        Object {
          "isAuthenticated": true,
        }
      }
      component={[Function]}
      exact={true}
    >
      <Route
        exact={true}
        render={[Function]}
      >
        <component
          history={
            Object {
              "action": "POP",
              "block": [Function],
              "createHref": [Function],
              "go": [Function],
              "goBack": [Function],
              "goForward": [Function],
              "listen": [Function],
              "location": Object {
                "hash": "",
                "pathname": "/",
                "search": "",
                "state": undefined,
              },
              "push": [Function],
              "replace": [Function],
            }
          }
          location={
            Object {
              "hash": "",
              "pathname": "/",
              "search": "",
              "state": undefined,
            }
          }
          match={
            Object {
              "isExact": true,
              "params": Object {},
              "path": "/",
              "url": "/",
            }
          }
          staticContext={Object {}}
        >
          <h1>
            Hello!
          </h1>
        </component>
      </Route>
    </privateRoute>
  </Router>
</StaticRouter>
`;
